### PR TITLE
NGWMN-1748 add messages describing conditions of the submitted data.

### PR DESCRIPTION
**Pull request by duselmann**

### CODE REVIEW

File: src/main/java/gov/usgs/wma/statistics/app/Application.java [hunk 1]
```diff
@@ -3,11 +3,20 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
 
-
+//@Configuration
```
Comment (ssoperMadisonCollege) [inline src/main/java/gov/usgs/wma/statistics/app/Application.java]: Remove if not needed.

File: src/main/java/gov/usgs/wma/statistics/app/Application.java [hunk 2]
```diff
@@ -3,11 +3,20 @@
...
 public class Application extends SpringBootServletInitializer {
+	
     public static void main(String[] args) {
         SpringApplication.run(Application.class, args);
     }
+    
+//	@Bean
+//	public Properties instance() {
+//		return new Properties();
+//	}
```
Comment (ssoperMadisonCollege) [inline src/main/java/gov/usgs/wma/statistics/app/Application.java]: Remove if not needed

File: src/main/java/gov/usgs/wma/statistics/app/Properties.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	public static final String ENV_INVALID_ROW_DATE_FUTURE  = "app.error.invalid.row.date.future";
+	
+	public static final String ENV_MESSAGE_PROVISIONAL_RULE = "app.message.provisional.rule";
+	public static final String ENV_MESSAGE_MONTHLY_RULE     = "app.message.monthly.rule";
+	public static final String ENV_MESSAGE_MONTHLY_DETAIL   = "app.message.monthly.detail";
+	public static final String ENV_MESSAGE_MONTHLY_MEDIANS  = "app.message.monthly.medians";
+	public static final String ENV_MESSAGE_DATE_FIX_DAY     = "app.message.date.fix.day";
+	public static final String ENV_MESSAGE_DATE_FIX_MONTH   = "app.message.date.fix.month";
+	public static final String ENV_MESSAGE_OMIT_NULL        = "app.message.omit.null";
+	public static final String ENV_MESSAGE_OMIT_PROVISIONAL = "app.message.omit.provisional";
```
Comment (ssoperMadisonCollege) [inline src/main/java/gov/usgs/wma/statistics/app/Properties.java]: Neat idea for error messaging!

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java [hunk 4]
```diff
@@ -66,86 +77,131 @@ public JsonData calculate(
...
+				builder.mediation(mediationType);
+			} catch (Exception e) {
+				String validMediations = MediationType.validMediations();
+				String msg = env.getError(ENV_INVALID_MEDIATION, mediation, validMediations);
+				builder.error(msg);
 			}
-			
+			// parse medians param
+			if (isNotBlank(medians)) {
+				if ( ! (BOOLEAN_FALSE.equalsIgnoreCase(medians) || BOOLEAN_TRUE.equalsIgnoreCase(medians)) ) {
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: Could maybe split line 117-120 out into its own method.

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java [hunk 5]
```diff
@@ -66,86 +77,131 @@ public JsonData calculate(
...
 				
+				BigDecimal value = new BigDecimal(val);
 				WLSample sample = new WLSample(time, value, "ft", value, "", true, "", value);
 				
 				if (cols.length == 3) {
-					sample.setProvsional(true);
+					// Trim to remove whitespace and substring to convert full words like Provisional
+					String aging = cols[2].trim().substring(0, 1);
+					
+					if (Value.PROVISIONAL_CODE.equalsIgnoreCase(aging)) {
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: Could split lines 188-195 into its own method to reduce nested if logic.

File: src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		super(value);
+		this.sigfigs = sigfigs;
+	}
+
+	public ScientificDecimal(String value) {
+		super(value);
+		sigfigRules(value);
+	}
+	
+	 // TODO the proper handling of 0.000... will need to be revisited if my "guess" is incorrect.
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java]: Is 0.000 handled properly? If so feel free to remove this comment.

File: src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java [hunk 7]
```diff
@@ -86,10 +93,44 @@ public JsonData calculate(Specifier spec, List<S> samples)  {
...
+			} else if ( today.compareTo(utc) == -1 ) {
+				msg = env.getError(ENV_INVALID_ROW_DATE_FUTURE, i+1, utc);
+				builder.error(msg);
+			} else {
+				String fixed = fixMissingMonthAndDay(utc);
+				int delta = fixed.length() - utc.length();
+				sample.time = fixed;
+				if (delta <= 0) {
+					continue; // date is fine and was not fixed
+				} else if (delta <= 3) {
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java]: Would the condition on line 120 and the condition on line 122 both be satisfied if the delta was less than or equal to 0?  Is that intentional?

File: src/test/java/gov/usgs/wma/statistics/logic/ValdiationMessagesTest.java [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+			val = new BigDecimal(value);
+		}
+		return new WLSample(time, val, "units", val, "comment", true, "pcode", null);
+	}
+	private WLSample createSample(String time, String value, boolean provisional) {
+		WLSample sample = createSample(time, value);
+		sample.setProvsional(provisional);
+		return sample;
+	}
+	private Optional<String> fetchMessageLike(String property, int length) {
```
Comment (ssoper-usgs) [inline src/test/java/gov/usgs/wma/statistics/logic/ValdiationMessagesTest.java]: Neat helper method.

File: src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java (hunk 6 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java]: We do not have direction for how to handle this yet.

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java (hunk 5 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: yep

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java (hunk 4 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: yep

File: src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java (hunk 7 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java]: First, check if it is <= Zero and continue.
Now, it must be >= 1 but we want to handle the between 1 and 3 inclusive.

File: src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java (hunk 6 above)
Comment (ssoper-usgs) [reply src/main/java/gov/usgs/wma/statistics/logic/ScientificDecimal.java]: Roger that.

File: src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java (hunk 7 above)
Comment (ssoper-usgs) [reply src/main/java/gov/usgs/wma/statistics/logic/StatisticsCalculator.java]: Gotcha, that makes sense!

**Pull request merged**